### PR TITLE
This is the second step in refactoring the libsrc4 code.

### PR DESCRIPTION
**Pull request by DennisHeimbigner**
The first was branch newhash0.dmh.
As with newhash0.dmh, these changes should be transparent.
Primary changes are to add a new indexing type NCindex.

I am suspending further work on rebuilding libsrc4 both because
it is a larger job than I originally thought and also because there are
some higher priority tasks I need to attend to.

### CODE REVIEW

**Comment by WardF** [comment]
This pull request **introduces 2 alerts** when merging 8cb1fc4cfeea9eca3c37abed4794ad0356897275 into fc6ab9881f285f45d406fe07d906a63f5d02f713 - [view on lgtm.com](https://lgtm.com/projects/g/Unidata/netcdf-c/rev/pr-452769869dd7674c0b7fcca723bf070514613af9)

**new alerts:**

* 2 for Unsigned comparison to zero

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by edhartnett** [comment]
@DennisHeimbigner when I run this branch with address sanitizer, unfortunately seeing some memory errors.

Here's what I do:

```
git checkout newhash1.dmh
autoreconf -i && CFLAGS='-Wall -g -fsanitize=address -fno-omit-frame-pointer' CPPFLAGS='-I/usr/local/hdf5-1.10.1/include' LDFLAGS='-L/usr/local/hdf5-1.10.1/lib' ./configure --disable-utilities --disable-dap && make -j check
```

I get failures in nctest. When I go to that directory and try tst_rename, I get these errors:

```
ed@mikado:~/unidata/netcdf-c/nctest$ ./tst_rename

*** Testing v3/v4 API versions of some v2 tests.
*** testing simple dim rename...ok.
*** testing dim rename from nctest...ok.
*** Tests successful!

=================================================================
==17640==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 13 byte(s) in 1 object(s) allocated from:
    #0 0x7f1aed882f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x7f1aed4201ae in nc_utf8proc_map_custom /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:727
    #2 0x7f1aed420056 in nc_utf8proc_map /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:701
    #3 0x7f1aed40351c in nc_utf8_normalize /home/ed/unidata/netcdf-c/libdispatch/dutf8.c:81
    #4 0x7f1aed478394 in NC_finddim /home/ed/unidata/netcdf-c/libsrc/dim.c:148
    #5 0x7f1aed479f9b in NC3_rename_dim /home/ed/unidata/netcdf-c/libsrc/dim.c:469
    #6 0x7f1aed3e5e7b in nc_rename_dim /home/ed/unidata/netcdf-c/libdispatch/ddim.c:287
    #7 0x55cefc79067a in main /home/ed/unidata/netcdf-c/nctest/tst_rename.c:252
    #8 0x7f1aecfe31c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)

Direct leak of 13 byte(s) in 1 object(s) allocated from:
    #0 0x7f1aed882f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x7f1aed4201ae in nc_utf8proc_map_custom /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:727
    #2 0x7f1aed420056 in nc_utf8proc_map /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:701
    #3 0x7f1aed40351c in nc_utf8_normalize /home/ed/unidata/netcdf-c/libdispatch/dutf8.c:81
    #4 0x7f1aed478394 in NC_finddim /home/ed/unidata/netcdf-c/libsrc/dim.c:148
    #5 0x7f1aed479f9b in NC3_rename_dim /home/ed/unidata/netcdf-c/libsrc/dim.c:469
    #6 0x7f1aed3e5e7b in nc_rename_dim /home/ed/unidata/netcdf-c/libdispatch/ddim.c:287
    #7 0x55cefc79079f in main /home/ed/unidata/netcdf-c/nctest/tst_rename.c:253
    #8 0x7f1aecfe31c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)

Direct leak of 9 byte(s) in 1 object(s) allocated from:
    #0 0x7f1aed882f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x7f1aed4201ae in nc_utf8proc_map_custom /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:727
    #2 0x7f1aed420056 in nc_utf8proc_map /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:701
    #3 0x7f1aed40351c in nc_utf8_normalize /home/ed/unidata/netcdf-c/libdispatch/dutf8.c:81
    #4 0x7f1aed478394 in NC_finddim /home/ed/unidata/netcdf-c/libsrc/dim.c:148
    #5 0x7f1aed479f9b in NC3_rename_dim /home/ed/unidata/netcdf-c/libsrc/dim.c:469
    #6 0x7f1aed3e5e7b in nc_rename_dim /home/ed/unidata/netcdf-c/libdispatch/ddim.c:287
    #7 0x55cefc7901b2 in main /home/ed/unidata/netcdf-c/nctest/tst_rename.c:248
    #8 0x7f1aecfe31c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)
...
```

nctest also fails.

When I run make check in nc_test4 I see only a handful of failures so it appears that the leaks are at least not too widespread.

```
FAIL: tst_dims
PASS: tst_dims2
PASS: tst_dims3
FAIL: tst_files
PASS: tst_files4
FAIL: tst_vars
FAIL: tst_varms
PASS: tst_unlim_vars
PASS: tst_converts
PASS: tst_converts2
PASS: tst_grps
PASS: tst_grps2
PASS: tst_compounds
PASS: tst_compounds2
PASS: tst_compounds3
PASS: tst_opaques
PASS: tst_strings
PASS: tst_strings2
PASS: tst_interops
PASS: tst_interops4
PASS: tst_interops5
PASS: tst_interops6
PASS: tst_enums
PASS: tst_coords
PASS: tst_coords2
PASS: tst_coords3
PASS: tst_vars3
PASS: tst_vars4
PASS: tst_chunks
PASS: tst_chunks2
FAIL: tst_utf8
PASS: tst_fills
PASS: tst_fills2
PASS: tst_fillbug
PASS: tst_xplatform
PASS: tst_xplatform2
PASS: tst_endian_fill
PASS: tst_atts
PASS: t_type
PASS: cdm_sea_soundings
PASS: tst_camrun
PASS: tst_vl
PASS: tst_atts1
PASS: tst_atts2
FAIL: tst_vars2
PASS: tst_files5
PASS: tst_files6
PASS: tst_sync
PASS: tst_h_scalar
PASS: tst_rename
PASS: tst_rename2
PASS: tst_h5_endians
PASS: tst_atts_string_rewrite
PASS: tst_hdf5_file_compat
PASS: tst_fill_attr_vanish
FAIL: tst_rehash
PASS: tst_filterparser
PASS: tst_bug324
PASS: tst_types
PASS: tst_h_strbug
PASS: tst_h_refs
PASS: run_empty_vlen_test.sh
FAIL: tst_v2
```

**Comment by DennisHeimbigner** [comment]
Thanks. I currently do not have access to a working leak detector.
Anyway I will fix shortly.

**Comment by edhartnett** [comment]
@DennisHeimbigner not sure if you are aware but there are still leaks. I didn't actually post all the error messages before, just some of them. So your fixes fixed some but left others.

You can use the gcc memory checker just by setting CFLAGS:

`CFLAGS='-Wall -g -fsanitize=address -fno-omit-frame-pointer'
`

Here's some of the memory errors I'm still getting on your branch.
```
FAIL: tst_rename
FAIL: nctest
PASS: compare_test_files.sh
============================================================================
Testsuite summary for netCDF 4.6.1-development
============================================================================
# TOTAL: 3
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  2
# XPASS: 0
# ERROR: 0
============================================================================
See nctest/test-suite.log
Please report to support-netcdf@unidata.ucar.edu
============================================================================
Makefile:795: recipe for target 'test-suite.log' failed
make[3]: *** [test-suite.log] Error 1
make[3]: Leaving directory '/home/ed/unidata/netcdf-c/nctest'
Makefile:901: recipe for target 'check-TESTS' failed
make[2]: *** [check-TESTS] Error 2
make[2]: Leaving directory '/home/ed/unidata/netcdf-c/nctest'
Makefile:981: recipe for target 'check-am' failed
make[1]: *** [check-am] Error 2
make[1]: Leaving directory '/home/ed/unidata/netcdf-c/nctest'
Makefile:671: recipe for target 'check-recursive' failed
make: *** [check-recursive] Error 1
ed@mikado:~/unidata/netcdf-c$ cd nctest
ed@mikado:~/unidata/netcdf-c/nctest$ ./tst_rename

*** Testing v3/v4 API versions of some v2 tests.
*** testing simple dim rename...ok.
*** testing dim rename from nctest...ok.
*** Tests successful!

=================================================================
==24513==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 3 byte(s) in 1 object(s) allocated from:
    #0 0x7f45c3320f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x7f45c2ebe1ae in nc_utf8proc_map_custom /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:727
    #2 0x7f45c2ebe056 in nc_utf8proc_map /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:701
    #3 0x7f45c2ea151c in nc_utf8_normalize /home/ed/unidata/netcdf-c/libdispatch/dutf8.c:81
    #4 0x7f45c2f13353 in NC_findvar /home/ed/unidata/netcdf-c/libsrc/var.c:374
    #5 0x7f45c2f1418d in NC3_def_var /home/ed/unidata/netcdf-c/libsrc/var.c:607
    #6 0x7f45c2e88f5e in nc_def_var /home/ed/unidata/netcdf-c/libdispatch/dvar.c:218
    #7 0x56025a829e70 in create_file /home/ed/unidata/netcdf-c/nctest/tst_rename.c:114
    #8 0x56025a82bbf0 in main /home/ed/unidata/netcdf-c/nctest/tst_rename.c:241
    #9 0x7f45c2a811c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)

Direct leak of 3 byte(s) in 1 object(s) allocated from:
    #0 0x7f45c3320f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x7f45c2ebe1ae in nc_utf8proc_map_custom /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:727
    #2 0x7f45c2ebe056 in nc_utf8proc_map /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:701
    #3 0x7f45c2ea151c in nc_utf8_normalize /home/ed/unidata/netcdf-c/libdispatch/dutf8.c:81
    #4 0x7f45c2f13353 in NC_findvar /home/ed/unidata/netcdf-c/libsrc/var.c:374
    #5 0x7f45c2f1418d in NC3_def_var /home/ed/unidata/netcdf-c/libsrc/var.c:607
    #6 0x7f45c2e88f5e in nc_def_var /home/ed/unidata/netcdf-c/libdispatch/dvar.c:218
    #7 0x56025a829fe1 in create_file /home/ed/unidata/netcdf-c/nctest/tst_rename.c:119
    #8 0x56025a82bbf0 in main /home/ed/unidata/netcdf-c/nctest/tst_rename.c:241
    #9 0x7f45c2a811c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)

Direct leak of 3 byte(s) in 1 object(s) allocated from:
    #0 0x7f45c3320f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x7f45c2ebe1ae in nc_utf8proc_map_custom /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:727
    #2 0x7f45c2ebe056 in nc_utf8proc_map /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:701
    #3 0x7f45c2ea151c in nc_utf8_normalize /home/ed/unidata/netcdf-c/libdispatch/dutf8.c:81
    #4 0x7f45c2f13353 in NC_findvar /home/ed/unidata/netcdf-c/libsrc/var.c:374
    #5 0x7f45c2f1418d in NC3_def_var /home/ed/unidata/netcdf-c/libsrc/var.c:607
    #6 0x7f45c2e88f5e in nc_def_var /home/ed/unidata/netcdf-c/libdispatch/dvar.c:218
    #7 0x56025a82a1c7 in create_file /home/ed/unidata/netcdf-c/nctest/tst_rename.c:125
    #8 0x56025a82bbf0 in main /home/ed/unidata/netcdf-c/nctest/tst_rename.c:241
    #9 0x7f45c2a811c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)

Direct leak of 3 byte(s) in 1 object(s) allocated from:
    #0 0x7f45c3320f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x7f45c2ebe1ae in nc_utf8proc_map_custom /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:727
    #2 0x7f45c2ebe056 in nc_utf8proc_map /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:701
    #3 0x7f45c2ea151c in nc_utf8_normalize /home/ed/unidata/netcdf-c/libdispatch/dutf8.c:81
    #4 0x7f45c2f13353 in NC_findvar /home/ed/unidata/netcdf-c/libsrc/var.c:374
    #5 0x7f45c2f1418d in NC3_def_var /home/ed/unidata/netcdf-c/libsrc/var.c:607
    #6 0x7f45c2e88f5e in nc_def_var /home/ed/unidata/netcdf-c/libdispatch/dvar.c:218
    #7 0x56025a82a2bb in create_file /home/ed/unidata/netcdf-c/nctest/tst_rename.c:129
    #8 0x56025a82bbf0 in main /home/ed/unidata/netcdf-c/nctest/tst_rename.c:241
    #9 0x7f45c2a811c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)

Direct leak of 3 byte(s) in 1 object(s) allocated from:
    #0 0x7f45c3320f40 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdef40)
    #1 0x7f45c2ebe1ae in nc_utf8proc_map_custom /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:727
    #2 0x7f45c2ebe056 in nc_utf8proc_map /home/ed/unidata/netcdf-c/libdispatch/utf8proc.c:701
    #3 0x7f45c2ea151c in nc_utf8_normalize /home/ed/unidata/netcdf-c/libdispatch/dutf8.c:81
    #4 0x7f45c2f13353 in NC_findvar /home/ed/unidata/netcdf-c/libsrc/var.c:374
    #5 0x7f45c2f1418d in NC3_def_var /home/ed/unidata/netcdf-c/libsrc/var.c:607
    #6 0x7f45c2e88f5e in nc_def_var /home/ed/unidata/netcdf-c/libdispatch/dvar.c:218
    #7 0x56025a829d7c in create_file /home/ed/unidata/netcdf-c/nctest/tst_rename.c:110
    #8 0x56025a82bbf0 in main /home/ed/unidata/netcdf-c/nctest/tst_rename.c:241
    #9 0x7f45c2a811c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)

SUMMARY: AddressSanitizer: 15 byte(s) leaked in 5 allocation(s).
```

**Comment by DennisHeimbigner** [comment]
Unfortunately, I cannot use valgrind on cygwin and I have so far
failed to get the gcc memory checker to work on cygwin. So
I have no way to fix the memory errors at the moment.

**Comment by WardF** [comment]
@DennisHeimbigner I was able to fix one of the issues observed by the memory sanitizer; one fix cleared up a lot of the failures, I suspect fixing the issue below will solve the few that are left:

````{.bash}
wfisher@kashian:~/Desktop/gitprojects/netcdf-c/build/ncdump$ ./tst_netcdf4.sh 

*** Testing ncgen and ncdump for netCDF-4 format.
*** Creating test output tst_netcdf4_c0.nc.
*** Testing that program tst_h_rdc0 can read tst_netcdf4_c0.nc.

*** Checking HDF5 file c0.nc.
*** Checking HDF5 objcts...ok.
*** Tests successful!
*** Running tst_create_files.c to create test files.

*** Creating test files for ncdump.
*** creating nested group file tst_solar_1.nc...ok.
*** checking nested group file tst_solar_1.nc...ok.
*** creating file with VLEN tst_solar_2.nc...ok.
*** creating file with compound type tst_solar_cmp.nc...ok.
*** Tests successful!
*** Testing tst_create_files output for netCDF-4 features.
*** Running tst_group_data.c to create test files.

*** Testing groups.
*** creating nested group file tst_group_data.nc...ok.
*** Tests successful!
*** Testing -v option with absolute name and groups...
*** Testing -v option with relative name and groups...
*** Running tst_enum_data.c to create test files.

*** Testing enums.
*** creating enum test file tst_enum_data.nc...ok.
*** Tests successful!
*** Running tst_opaque_data.c to create test files.

*** Testing opaque types.
*** creating opaque test file tst_opaque_data.nc...ok.
*** Tests successful!
*** Running tst_vlen_data.c to create test files.

*** Testing vlens.
=================================================================
==58348==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffee10c0ef4 at pc 0x00010f31ed78 bp 0x7ffee10bffb0 sp 0x7ffee10bf760
READ of size 16 at 0x7ffee10c0ef4 thread T0
    #0 0x10f31ed77 in __asan_memcpy (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x50d77)
    #1 0x10ec969bf in NC4_put_att nc4attr.c:832
    #2 0x10eb5a752 in nc_put_att dattput.c:235
    #3 0x10eb41536 in main tst_vlen_data.c:64
    #4 0x7fff56340114 in start (libdyld.dylib:x86_64+0x1114)

Address 0x7ffee10c0ef4 is located in stack of thread T0 at offset 628 in frame
    #0 0x10eb4082f in main tst_vlen_data.c:28

  This frame has 17 object(s):
    [32, 36) 'ncid' (line 29)
    [48, 52) 'dimid' (line 30)
    [64, 68) 'varid' (line 30)
    [80, 84) 'typeid' (line 31)
    [96, 353) 'name_in' (line 32)
    [432, 436) 'base_typeid' (line 33)
    [448, 456) 'base_size_in' (line 34)
    [480, 484) 'class_in' (line 35)
    [496, 500) 'var_dims' (line 40)
    [512, 592) 'ragged_data' (line 42)
    [624, 628) 'missing_value' (line 43)
    [640, 656) 'missing_val' (line 44) <== Memory access at offset 628 partially underflows this variable
    [672, 688) 'val_in' (line 45)
    [704, 712) 'index' (line 122)
    [736, 816) 'vals_in' (line 136)
    [848, 856) 'start' (line 138)
    [880, 888) 'count' (line 138)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x50d77) in __asan_memcpy
Shadow bytes around the buggy address:
  0x1fffdc218180: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdc218190: f1 f1 f1 f1 04 f2 04 f2 04 f2 04 f2 00 00 00 00
  0x1fffdc2181a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdc2181b0: 00 00 00 00 00 00 00 00 00 00 00 00 01 f2 f2 f2
  0x1fffdc2181c0: f2 f2 f2 f2 f2 f2 04 f2 00 f2 f2 f2 04 f2 04 f2
=>0x1fffdc2181d0: 00 00 00 00 00 00 00 00 00 00 f2 f2 f2 f2[04]f2
  0x1fffdc2181e0: 00 00 f2 f2 00 00 f2 f2 f8 f2 f2 f2 f8 f8 f8 f8
  0x1fffdc2181f0: f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f8 f2 f2 f2 f8 f3
  0x1fffdc218200: f3 f3 f3 f3 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdc218210: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdc218220: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==58348==ABORTING
*** creating vlen test file tst_vlen_data.nc..../tst_netcdf4.sh: line 62: 58348 Abort trap: 6           ${execdir}/tst_vlen_data
wfisher@kashian:~/Desktop/gitprojects/netcdf-c/build/ncdump$ 
````

The top level offending chunk of code is as follows:
````{.C}
     /* Copy the fill_value. */
      LOG((4, "Copying fill value into metadata for variable %s", var->name));
      if (var->type_info->nc_type_class == NC_VLEN)
      {
         nc_vlen_t *in_vlen = (nc_vlen_t *)data, *fv_vlen = (nc_vlen_t *)(var->fill_value);

         fv_vlen->len = in_vlen->len;
         if (!(fv_vlen->p = malloc(size * in_vlen->len)))
            return NC_ENOMEM;
         memcpy(fv_vlen->p, in_vlen->p, in_vlen->len * size);
      }
````

**Comment by DennisHeimbigner** [comment]
ok, I fixed a bunch more related to utf8_normalize in libsrc.
I will check them in shortly.

File: include/nclistmap.h [hunk 1]
```diff
@@ -7,13 +7,26 @@ See LICENSE.txt for license information.
 #define NCLISTMAP_H
 
 #include "nclist.h"
+<<<<<<< HEAD
```
Comment (WardF) [inline include/nclistmap.h]: It appears that a merge/diff was accepted without being resolved.  @DennisHeimbigner can you providie guidance as to how we should resolve this?

Comment (DennisHeimbigner) [reply include/nclistmap.h]: I think when I was resolving conflicts, I forget to remove that <<<<HEAD line. It should just be removed.

**Comment by edhartnett** [comment]
@WardF will this PR be merged soon? I would like to present the user-defined format and separate HDF5 code features, but this PR should be merged first, to avoid conflicts. Thanks.

**Comment by WardF** [comment]
We are seeing failures on Windows that I will have to debug.  I'm also under the gun on a deadline for an NSF proposal, but it will be merged ASAP.

**Comment by edhartnett** [comment]
OK, thanks. I will check in again next week. ;-)

**Comment by WardF** [comment]
Good news! The latest updates appear to have fixed the Windows issue.  I'll merge this later today when the CI finishes.

**Pull request merged**